pytorch_lightning/trainer/trainer.py

```diff
@@ -1046,10 +1046,16 @@ def _log_hyperparams(self):
 
                 colliding_keys = lightning_hparams.keys() & datamodule_hparams.keys()
                 if colliding_keys:
-                    raise MisconfigurationException(
-                        f"Error while merging hparams: the keys {colliding_keys} are present "
-                        "in both the LightningModule's and LightningDataModule's hparams."
-                    )
+                    inconsistent_keys = []
+                    for key in colliding_keys:
+                        if lightning_hparams[key] != datamodule_hparams[key]:
```

Comment:
here we assume that these objects can be compared with ==,  that's probably too strong.

For example, if these are tensors then the result of the comparison is again a tensor and we get an error. Or if these tensors don't even have the same shape, we yet get another error.

Reply:
is there a specification around what argument types can be included in the hparams, which could then guide these equality checks? is there another form of equality check we ought to use for this comparison?

Reply:
Should we use id(lightning_hparams[key]) != id(datamodule_hparams[key]) ?

Reply:
> is there a specification around what argument types can be included in the hparams


not really afaik. there are two functionalities:
1) saving hparams to checkpoint
2) logging hparams

In 1) all we require is that the object is picklable. 
In 2) it is ultimately the logger defining what it can log and what it can't log.

Reply:
@awaelchli I updated to use `id` in case this would be safer. i'm not sure if this introduces issues around lifetimes or hparams potentially being updated in between when save_hyperparameters is called and when it's logged here though.

Reply:
sounds good. hopefully nobody is using ints as keys hehe. See this [funny business here](https://github.com/satwikkansal/wtfpython#-how-not-to-use-is-operator). Where `id(257) != id(257)` but `id(256) == id(256)` 🤣

Reply:
If the problem is just tensors, we can use id only for them, instead of id for everything

Reply:
@carmocca @awaelchli I added special asing for tensors to use the id check. LMK if you think this is good to merge now, or if there's a more general approach